### PR TITLE
Allows passing through class weights to loss function from config

### PR DESCRIPTION
**Pull request by jdunnmon**

### CODE REVIEW

**Comment by bhancock8** [comment]
To be a total nit jerk here:

"A properly formed git commit subject line should always be able to complete the following sentence
*If applied, this commit will <your subject line here>*

Rules for a great git commit message style:
- Separate subject from body with a blank line
- Do not end the subject line with a period
- Capitalize the subject line and each paragraph
- Use the imperative mood in the subject line
- Wrap lines at 72 characters
- Use the body to explain what and why you have done something. In most cases, you can leave out details about how a change has been made."

File: metal/end_model/loss.py [hunk 1]
```diff
@@ -20,14 +20,19 @@ class SoftCrossEntropyLoss(nn.Module):
 
     def __init__(self, weight=None, reduction="elementwise_mean"):
         super().__init__()
-        assert weight is None or isinstance(weight, torch.FloatTensor)
         self.weight = weight
+        if self.weight is not None:
+            self.weight = torch.Tensor(self.weight)
```
Comment (ajratner) [inline metal/end_model/loss.py]: What's the rationale for not explicitly forming a `FloatTensor` here?  Wouldn't that be better?

File: metal/end_model/loss.py [hunk 2]
```diff
@@ -20,14 +20,19 @@ class SoftCrossEntropyLoss(nn.Module):
...
+        assert self.weight is None or isinstance(self.weight, torch.FloatTensor)
         self.reduction = reduction
 
     def forward(self, input, target):
         n, k = input.shape
         # Note that t.new_zeros, t.new_full put tensor on same device as t
         cum_losses = input.new_zeros(n)
+        # Moving weights to appropriate device
+        if self.weight is not None:
+            self.weight = self.weight.type(input.dtype).to(input.device)
```
Comment (ajratner) [inline metal/end_model/loss.py]: This seems clunky?  It seems like the torch-preferred way is to use [`register_buffer`](https://pytorch.org/docs/stable/nn.html#torch.nn.Module.register_buffer) here

File: metal/end_model/end_model.py [hunk 3]
```diff
@@ -81,8 +81,11 @@ def _build(self, input_module, middle_modules, head_module):
             self.network = layers[0]
 
         # Construct loss module
+        weight = self.config["train_config"]["loss_weights"]
+        if weight is not None:
```
Comment (ajratner) [inline metal/end_model/end_model.py]: Should add a verbosity check here

File: metal/end_model/loss.py [hunk 4]
```diff
@@ -20,8 +20,12 @@ class SoftCrossEntropyLoss(nn.Module):
...
     def __init__(self, weight=None, reduction="elementwise_mean"):
         super().__init__()
-        assert weight is None or isinstance(weight, torch.FloatTensor)
-        self.weight = weight
+        self.weight = None
+        if weight is not None:
+            self.weight = torch.FloatTensor(weight)
+            # Register as buffer is standard way to make sure gets moved /
+            # converted with the Module, without making it a Parameter
+            self.register_buffer("class_weights", self.weight)
```
Comment (bhancock8) [inline metal/end_model/loss.py]: @ajratner--nice, register_buffer is much cleaner! But it seems a little confusing to me that in our defaults it's 'loss_weights', in the module it's self.weight, and then we register it as 'class_weights'. I vote we make it 'loss_weights' everywhere.

**Comment by ajratner** [comment]
Sure sg by me!

**Comment by bhancock8** [comment]
Made the change in 7006faf. Note that register_buffer() also stores the attribute. So before, we couldn't do register_buffer() with the name "loss_weights" after we had already assigned self.loss_weights a couple lines before; it complains that "loss_weights" is already defined. No doubt that's why it was given a different name "class_weights". I think the right way to fix it (and what I did) is to move register_buffer up to the place where we define that attribute.

**Comment by ajratner** [comment]
Ah @bhancock8 good catch!
